### PR TITLE
[CURL][SOUP] Extract common data URL logic into shared NetworkDataTaskDataURL

### PR DESCRIPTION
**Pull request by yury-s**
#### afb1339d37f44414e98767bb0bc75cb0ab71c139
<pre>
[CURL][SOUP] Extract common data URL logic into shared NetworkDataTaskDataURL
<a href="https://bugs.webkit.org/show_bug.cgi?id=255750">https://bugs.webkit.org/show_bug.cgi?id=255750</a>

Reviewed by Fujii Hironori.

Curl and Soup don&apos;t support data: URLs and the download logic for both
of these platforms is mostly the same. This change extracts common bits
of data urls handling into NetworkDataTaskDataURL which can be shared
by both platforms.

* Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp: Added.
(WebKit::NetworkDataTaskDataURL::create):
(WebKit::NetworkDataTaskDataURL::NetworkDataTaskDataURL):
(WebKit::NetworkDataTaskDataURL::~NetworkDataTaskDataURL):
(WebKit::NetworkDataTaskDataURL::resume): note that this task never gets back
to the suspended state once it has been initially resumed.
(WebKit::NetworkDataTaskDataURL::cancel):
(WebKit::NetworkDataTaskDataURL::invalidateAndCancel):
(WebKit::NetworkDataTaskDataURL::state const):
(WebKit::NetworkDataTaskDataURL::setPendingDownloadLocation):
(WebKit::NetworkDataTaskDataURL::suggestedFilename const):
(WebKit::NetworkDataTaskDataURL::didDecodeDataURL):
(WebKit::NetworkDataTaskDataURL::downloadDecodedData):
* Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.h: Added.
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
(WebKit::NetworkDataTaskSoup::create):
(WebKit::NetworkDataTaskSoup::createRequest):
(WebKit::NetworkDataTaskSoup::clearRequest):
(WebKit::NetworkDataTaskSoup::resume):
(WebKit::NetworkDataTaskSoup::didReadDataURL): Deleted.
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:

Canonical link: <a href="https://commits.webkit.org/263349@main">https://commits.webkit.org/263349@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/45c759a19b7cabbed45dcda219ce0d3f7c48b805

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4035 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4246 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5481 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4286 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4229 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4110 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4508 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4087 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4276 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3643 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5464 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3619 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5810 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3602 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3680 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5180 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3291 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3603 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3619 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1065 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3656 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3881 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp
Comment (fujii) [inline Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp]: I think NetworkDataTask::create is better place to add this code. WDYT?
~~~
#if PLATFORM(COCOA)
    return NetworkDataTaskCocoa::create(session, client, parameters);
#else
    if (parameters.request.url().protocolIsData())
        return NetworkDataTaskDataURL::create(session, client, parameters);
~~~

https://github.com/WebKit/WebKit/blob/a0dcbd369295e3a93b80cbc22b0988678317c18f/Source/WebKit/NetworkProcess/NetworkDataTask.cpp#L58

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (fujii) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:33]: I think we should guard this file by "#if USE(CURL) && USE(SOUP)".

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.h
Comment (fujii) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.h:32]: I think we should guard this file by "#if USE(CURL) && USE(SOUP)".

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (fujii) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:84]: The original code is checking if `m_state == State::Suspended`. Do we need the check?
https://github.com/WebKit/WebKit/blob/a0dcbd369295e3a93b80cbc22b0988678317c18f/Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp#L349-L352

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (fujii) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:178]: What's the reason you want to use a variable `downloadPtr`. You want to avoid extra cost of `unique_ptr::operator->`?
I think typical WebKit code uses `download->`.

**Comment by fujii** [comment]
Cc: @kshukuwa

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (fujii) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:84]: ah, I found out your answer in the commit log.
~~~
(WebKit::NetworkDataTaskDataURL::resume): note that this task never gets back
to the suspended state once it has been initially resumed.
~~~

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (mcatanzaro) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:33]: `#if USE(CURL) || USE(SOUP)`, not `&&`, since they'll never both be true.

File: Source/WebKit/SourcesGTK.txt
Comment (kshukuwa) [inline Source/WebKit/SourcesGTK.txt]: Since the directory hierarchy is different, please change the insert location and leave a space for one line.

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (kshukuwa) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:82]: We have checked m_client for the SOUP port, but not checked here. Are there any issues going around?

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (kshukuwa) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp]: What is m_decodingState ?

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (kshukuwa) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp]: Is it possible to delete m_dataURLResult and pass it as an argument to functions ?

didDecodeDataURL(std::optional<WebCore::DataURLDecoder::Result>&& result) =>
didReceiveResponse([result = WTFMove(result)]) =>
downloadDecodedData(std::optional<WebCore::DataURLDecoder::Result>&& result)

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (kshukuwa) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:133]: Please change m_state to State::Completed after error occurs. This will prevent malfunctions.

Comment (kshukuwa) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:133]: You should call invalidateAndCancel() here.

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (kshukuwa) [inline Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:166]: CURLE_WRITE_ERROR does not represent the correct error. If there is no problem, why not return the same error as SOUP?

```
ResourceError(API::Error::webKitDownloadErrorDomain(), API::Error::Download::Destination, response.url(), localizedDescription);
```

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (kshukuwa) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:178]: I think this is because after doing WTFMove(download) you can no longer access download.

Comment (fujii) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:178]: Ah, existing code does same.
However, we have a rule to inhibit using a raw pointer of local variable. It'd nice we will have a solution for this problem.
https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines

File: Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp
Comment (yury-s) [reply Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp]: Done. I put the new task under USE(SOUP) for now as this is supposed to be a refactoring and I don't want to deal with possible windows issues in the same change. Will enable it for Windows in the separate [change](https://github.com/WebKit/WebKit/pull/12952)).

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (yury-s) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:84]: Yes,  I had suspended state in the first iteration of the change but then realized it was unnecessary. Avoiding extra code for handling the suspended state here simplifies the logic quite a bit.

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (yury-s) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:82]: None that I'm aware of, we check m_client later before we call it.

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (yury-s) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp]: Done.

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (yury-s) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp]: It's a leftover from the previous iteration where I was maintaining extra `suspended` state logic. Removed.

File: Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp
Comment (yury-s) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:166]: > CURLE_WRITE_ERROR does not represent the correct error.

Can you explain why you think so?

From the [curl documentation](https://curl.se/libcurl/c/libcurl-errors.html):
```
CURLE_WRITE_ERROR (23)

An error occurred when writing received data to a local file, or an error was returned to libcurl from a write callback.
```
Also it is the same error code that is [already used](https://github.com/WebKit/WebKit/blob/3f549dfda7946f5b554aa8c45da58e0fb97871aa/Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp#L203) to report similar download errors in webkit.

>  If there is no problem, why not return the same error as SOUP?

There is a problem, `downloadDestinationError` as well as `API::Error::webKitDownloadErrorDomain()` are only available in SOUP (guarded by `USE(SOUP)`).

Comment (kshukuwa) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:166]: > Can you explain why you think so?

Because it's not an error caused by curl. However, I understand that CURLE_WRITE_ERROR is already used in some places.

NetworkDataTaskBlob seems to return canceledError() in similar cases, or an error of its own. It's not a strong opinion, but why not consider these methods as well?

https://github.com/WebKit/WebKit/blob/29adb2b287fb17f11ff8f085c40b543b90166907/Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp#L439-L443

https://github.com/WebKit/WebKit/blob/29adb2b287fb17f11ff8f085c40b543b90166907/Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp#L528

https://github.com/WebKit/WebKit/blob/29adb2b287fb17f11ff8f085c40b543b90166907/Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h#L75-L82

Comment (yury-s) [reply Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:166]: In SOUP downloadDestinationError is used consistently for similar errors, so as far as I understand we are discussing only Curl behavior. In my opinion  `CURLE_WRITE_ERROR` more accurately reflects the gist of the error, but it's not a very strong opinion either.  I'd leave it to the code owners, if they prefer `cancelledError` I can change that (we'll likely need to update the other place as well).

@fujii what are your thoughts?

**Review by fujii — APPROVED** [review]
LGTM. Thank you very much.

**Pull request merged**
